### PR TITLE
SSL: Server & Client

### LINKED ISSUE #229: Unable to use zio-http as simple http client [Support request]
I’m trying to use zio-http as simple http client library with the SimpleClient example.

When I run the following code : 
```scala
object SimpleClient extends App {
  val env = ChannelFactory.auto ++ EventLoopGroup.auto()

  val program = for {
    _ <- console.putStrLn("Start")
    res <- Client.request("https://api.github.com/users/zio/repos")
    _ <- console.putStrLn("Finish")
    _ <- console.putStrLn {
      res.content match {
        case HttpData.CompleteData(data) => data.map(_.toChar).mkString
        case HttpData.StreamData(_)      => "<Chunked>"
        case HttpData.Empty              => ""
      }
    }
  } yield ()

  override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] = program.exitCode.provideCustomLayer(env)

}
```

Current Behavior:  When I run the sample code the request is not executed and the program is frozen
```
sbt:hello> run
[info] running SimpleClient 
Start
```

Expected Behavior:  When I run the sample code *res.content* is displayed

Sample code : https://github.com/jlcanela/zhttp-client

Any help welcome, I’m 99% sure I did something stupid but cannot find it at the moment.

### LINKED ISSUE #249: Server SSL Support

### PR DESCRIPTION
**Pull request by dinojohn**
closes #249 , closes  #229

### CODE REVIEW

File: example/src/main/scala/HttpsClient.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import zhttp.service.{ChannelFactory, Client, EventLoopGroup}
+import zio._
+
+import java.io.InputStream
+import java.security.KeyStore
+import javax.net.ssl.TrustManagerFactory
+
+object HttpsClient extends App {
+  val env     = ChannelFactory.auto ++ EventLoopGroup.auto()
+  val url     = "https://sports.api.decathlon.com/groups/water-aerobics"
```
Comment (tusharmath) [inline example/src/main/scala/HttpsClient.scala:13]: Use the GitHub API

File: example/src/main/scala/HttpsClient.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import io.netty.handler.ssl.SslContextBuilder
```
Comment (tusharmath) [inline example/src/main/scala/HttpsClient.scala:1]: Wrap the netty SSL builder using a custom one.

File: example/src/main/scala/HttpsHelloWorld.scala [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import io.netty.handler.ssl.ApplicationProtocolConfig.{
+  Protocol,
+  SelectedListenerFailureBehavior,
+  SelectorFailureBehavior,
+}
+import io.netty.handler.ssl.{ApplicationProtocolConfig, ApplicationProtocolNames, SslContextBuilder, SslProvider}
```
Comment (tusharmath) [inline example/src/main/scala/HttpsHelloWorld.scala]: Remove `io.netty` usage.

File: example/src/main/scala/HttpsHelloWorld.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    .sslProvider(SslProvider.JDK)
+    .applicationProtocolConfig(
+      new ApplicationProtocolConfig(
+        Protocol.ALPN,
+        SelectorFailureBehavior.NO_ADVERTISE,
+        SelectedListenerFailureBehavior.ACCEPT,
+        ApplicationProtocolNames.HTTP_1_1,
+      ),
+    )
+    .build()
```
Comment (tusharmath) [inline example/src/main/scala/HttpsHelloWorld.scala]: Create two separate examples — 
1. Default SSL settings to quickly start the server over SSL.
2. Showcase extensive customization with SSL.

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 5]
```diff
@@ -44,11 +48,13 @@ object Server {
   private final case class MaxRequestSize(size: Int)                                  extends UServer
   private final case class App[R, E](http: HttpApp[R, E])                             extends Server[R, E]
   private final case class Error[R](errorHandler: Throwable => ZIO[R, Nothing, Unit]) extends Server[R, Nothing]
+  private final case class Ssl(sslOptions: ServerSSLOptions)                          extends UServer
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala:51]:
```suggestion
  private final case class Secure(sslOptions: ServerSSLOptions)                          extends UServer
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 6]
```diff
@@ -44,11 +48,13 @@ object Server {
...
 
   def app[R, E](http: HttpApp[R, E]): Server[R, E]                                   = Server.App(http)
   def maxRequestSize(size: Int): UServer                                             = Server.MaxRequestSize(size)
   def port(int: Int): UServer                                                        = Server.Port(int)
   def error[R](errorHandler: Throwable => ZIO[R, Nothing, Unit]): Server[R, Nothing] = Server.Error(errorHandler)
+  def ssl(sslOptions: ServerSSLOptions): UServer                                     = Server.Ssl(sslOptions)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala:57]:
```suggestion
  def secure(sslOptions: ServerSSLOptions): UServer                                     = Server.secure(sslOptions)
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 7]
```diff
@@ -57,8 +63,12 @@ object Server {
   /**
    * Launches the app on the provided port.
    */
-  def start[R <: Has[_]](port: Int, http: RHttpApp[R]): ZIO[R, Throwable, Nothing] =
-    (Server.port(port) ++ Server.app(http)).make.useForever
+  def start[R <: Has[_]](
+    port: Int,
+    http: RHttpApp[R],
+    sslOptions: ServerSSLOptions = NoSSL,
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]: Remove `sslOptions` from the `start` method.

File: zio-http/src/main/scala/zhttp/service/client/ClientChannelInitializer.scala [hunk 8]
```diff
@@ -1,14 +1,25 @@
 package zhttp.service.client
 
+import io.netty.channel.ChannelPipeline
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/client/ClientChannelInitializer.scala]:
```suggestion
import io.netty.channel.{ChannelPipeline => JChannelPipeline}
```

File: zio-http/src/main/scala/zhttp/service/client/ClientChannelInitializer.scala [hunk 9]
```diff
@@ -1,14 +1,25 @@
...
 import io.netty.handler.codec.http.{HttpClientCodec => JHttpClientCodec}
+import io.netty.handler.ssl.SslContext
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/client/ClientChannelInitializer.scala]:
```suggestion
import io.netty.handler.ssl.{SslContext => JSslContext}
```

File: zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala [hunk 10]
```diff
@@ -2,20 +2,29 @@ package zhttp.service.server
...
+    val p = channel
       .pipeline()
       .addLast(SERVER_CODEC_HANDLER, new JHttpServerCodec)
       .addLast(HTTP_KEEPALIVE_HANDLER, new HttpServerKeepAliveHandler)
-      .addLast(OBJECT_AGGREGATOR, new JHttpObjectAggregator(maxSize))
+      .addLast(OBJECT_AGGREGATOR, new JHttpObjectAggregator(settings.maxRequestSize))
       .addLast(HTTP_REQUEST_HANDLER, httpH)
-    ()
+
+    ServerSSLHandler.ssl(settings.sslOption) match {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala]: potential perf issues with an `Option` allocation. Consider using `null`.

File: example/src/main/scala/HttpsClient.scala (hunk 2 above)
Comment (ShrutiVerma97) [reply example/src/main/scala/HttpsClient.scala:1]: Discussed already. There are n apis for creating `truststore` and `keystore` using SslContextBuilder.

File: example/src/main/scala/HttpsHelloWorld.scala [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  val app = HttpApp.collect {
+    case Method.GET -> Root / "text" => Response.text("Hello World!")
+    case Method.GET -> Root / "json" => Response.jsonString("""{"greetings": "Hello World!"}""")
+  }
+
+  /**
+   * custom sslcontext can be created using SslContexBuilder. In this example an inbuilt API using keystore is used
+   */
+  private val server = Server.port(8090) ++ Server.app(app) ++ Server.ssl(
+    SSLFromKeystore(getClass.getResourceAsStream("mysslstore.jks"), "123456", "123456"),
```
Comment (tusharmath) [inline example/src/main/scala/HttpsHelloWorld.scala]:
```suggestion
    SSLFromKeystore(getClass.getResourceAsStream("KeyStore.jks"), "ZIOHttp", "password"),
```

File: zio-http/src/test/scala/zhttp/service/SSLSpec.scala [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+          testM("fail with SSLHandshakeException when client doesn't have the server certificate") {
+            val actual = Client
+              .request("https://localhost:8080/success", ClientSSLOptions.CustomSSL(clientssl2))
+              .map(_.status)
+              .catchSome(_.getCause match {
+                case _: SSLHandshakeException => ZIO.succeed("SSLHandshakeException")
+              })
+            assertM(actual)(equalTo("SSLHandshakeException"))
+          },
+          testM("empty response when client makes http request") {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/SSLSpec.scala]: HTTP request should fail, if the server doesn't support HTTP.
Can add a failing test and fix in a separate PR.

File: zio-http/src/test/scala/zhttp/service/SSLSpec.scala [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+                .fork
+              _     <- TestClock.adjust(3 seconds)
+              res   <- fiber.join
+            } yield res
+            assertM(actual)(isNone)
+          },
+        ),
+      )
+      .useNow,
+  ).provideCustomLayer(env)
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/SSLSpec.scala:85]: Add test for client ignoring the SSL certificate and resolving successfully.

File: zio-http/src/test/scala/zhttp/service/SSLSpec.scala [hunk 14]
```diff
@@ -0,0 +1,83 @@
+package zhttp.service
+
+import io.netty.handler.ssl.ApplicationProtocolConfig.{
+  Protocol,
+  SelectedListenerFailureBehavior,
+  SelectorFailureBehavior,
+}
+import io.netty.handler.ssl.util.SelfSignedCertificate
+import io.netty.handler.ssl.{ApplicationProtocolConfig, ApplicationProtocolNames, SslContextBuilder, SslProvider}
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/SSLSpec.scala]: Prefix `J` with netty/java imports.

File: zio-http/src/test/scala/zhttp/service/HttpRunnableSpec.scala [hunk 15]
```diff
@@ -3,28 +3,35 @@ package zhttp.service
...
 import zio.{Chunk, Has, ZIO, ZManaged}
 
 abstract class HttpRunnableSpec(port: Int) extends DefaultRunnableSpec {
 
   def serve[R <: Has[_]](
     app: RHttpApp[R],
+    sslServerOptions: ServerSSLOptions = NoSSL,
   ): ZManaged[R with EventLoopGroup with ServerChannelFactory, Nothing, Unit] =
-    Server.make(Server.app(app) ++ Server.port(port)).orDie
+    Server.make(Server.app(app) ++ Server.port(port) ++ Server.ssl(sslServerOptions)).orDie
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/HttpRunnableSpec.scala]: Remove ssl options if SSL is not enabled.

File: zio-http/src/test/scala/zhttp/service/HttpRunnableSpec.scala [hunk 16]
```diff
@@ -3,28 +3,35 @@ package zhttp.service
...
 
   def status(path: Path): ZIO[EventLoopGroup with ChannelFactory, Throwable, Status] =
     requestPath(path).map(_.status)
 
   def requestPath(path: Path): ZIO[EventLoopGroup with ChannelFactory, Throwable, UHttpResponse] =
-    Client.request(Method.GET -> URL(path, Location.Absolute(Scheme.HTTP, "localhost", port)))
+    Client.request(
+      Method.GET -> URL(path, Location.Absolute(Scheme.HTTP, "localhost", port)),
+      ClientSSLOptions.DefaultSSL,
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/HttpRunnableSpec.scala:23]: Second argument should be optional.

File: zio-http/src/test/scala/zhttp/service/ClientHttpsSpec.scala [hunk 17]
```diff
@@ -0,0 +1,47 @@
+package zhttp.service
+
+import io.netty.handler.codec.DecoderException
+import io.netty.handler.ssl.SslContextBuilder
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ClientHttpsSpec.scala]: Use `J` prefix for Java classes.

File: zio-http/src/main/scala/zhttp/service/client/ClientChannelInitializer.scala [hunk 18]
```diff
@@ -1,14 +1,25 @@
...
   override def initChannel(ch: JChannel): Unit = {
-    ch.pipeline()
+    val sslCtx: JSslContext =
+      if (scheme == "https") ClientSSLHandler.ssl(sslOption) else null
+    val p: JChannelPipeline = ch
+      .pipeline()
       .addLast(new JHttpClientCodec)
       .addLast(new JHttpObjectAggregator(Int.MaxValue))
       .addLast(channelHandler)
+    if (sslCtx != null) p.addFirst(sslCtx.newHandler(ch.alloc))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/client/ClientChannelInitializer.scala]: can move adding of sslCtx.newHandler above with the creation of the JSslContext.

File: zio-http/src/test/scala/zhttp/service/SSLSpec.scala (hunk 12 above)
Comment (dinojohn) [reply zio-http/src/test/scala/zhttp/service/SSLSpec.scala]: it is made to redirect or accept http request

File: zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala [hunk 19]
```diff
@@ -1,21 +1,81 @@
 package zhttp.service.server
 
 import io.netty.handler.codec.http.HttpServerKeepAliveHandler
+import io.netty.handler.ssl.{OptionalSslHandler, SslHandler}
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala]:
```suggestion
import io.netty.handler.ssl.{OptionalSslHandler => JOptionalSslHandler, SslHandler => JSslHandler}
```

**Review by tusharmath — APPROVED** [review]

**Pull request merged**